--- main.c.org	Fri Mar 30 12:20:33 2001
+++ main.c	Fri Mar 30 12:29:01 2001
@@ -59,8 +59,10 @@
 int w3m_dump = 0;
 int w3m_dump_source = 0;
 int w3m_dump_head = 0;
+int w3m_dump_extra = 0;
 static void dump_source(Buffer *);
 static void dump_head(Buffer *);
+static void dump_extra(Buffer *);
 int prec_num = 0;
 int prev_key = -1;
 int on_target = 1;
@@ -270,6 +272,16 @@
 		if (COLS == 0)
 		    COLS = 80;
 	    }
+	    else if (!strcmp("-dump_extra", argv[i])) {
+		argv[i] = "-dummy";
+		w3m_dump = TRUE;
+		w3m_dump_head = TRUE;
+		w3m_dump_source = TRUE;
+		w3m_dump_extra = TRUE;
+		w3m_halfdump = FALSE;
+		if (COLS == 0)
+		    COLS = 80;
+	    }
 	    else if (!strcmp("-halfdump", argv[i])) {
 		argv[i] = "-dummy";
 		w3m_halfdump = TRUE;
@@ -659,6 +671,9 @@
 	    Currentbuf = newbuf;
 	}
 	if (w3m_dump) {
+	    if (w3m_dump_extra) {
+		dump_extra(Currentbuf);
+	    }
 	    if (w3m_dump_head)
 		dump_head(Currentbuf);
 	    if (w3m_dump_source)
@@ -793,6 +808,17 @@
 	printf("%s", ti->ptr);
     }
     puts("");
+}
+
+static void
+dump_extra(Buffer * buf)
+{
+    printf("W3m-current-url: %s\n",parsedURL2Str(&buf->currentURL)->ptr);
+    if (buf->baseURL)
+	printf("W3m-base-url: %s\n",parsedURL2Str(buf->baseURL)->ptr);
+#ifdef JP_CHARSET
+    printf("W3m-document-charset: %s\n", code_to_str(buf->document_code));
+#endif
 }
 
 void

