diff -cr w3m-0.2.1-inu-1.5.orig/configure w3m-0.2.1-inu-1.5/configure
*** w3m-0.2.1-inu-1.5.orig/configure	Fri Sep 28 22:17:38 2001
--- w3m-0.2.1-inu-1.5/configure	Tue Nov  6 14:14:01 2001
***************
*** 1316,1321 ****
--- 1316,1338 ----
    srandom_flg="#undef HAVE_SRANDOM"
  fi
  
+ ####### getpassphrase
+ cat > _zmachdep.c << EOF
+ #include <unistd.h>
+ main()
+ {
+   getpassphrase("");
+ }
+ EOF
+ if $cc $cflags -o _zmachdep _zmachdep.c > /dev/null 2>&1
+ then
+   echo "You have getpassphrase()."
+   getpassphrase_flg="#define HAVE_GETPASSPHRASE"
+ else
+   echo "You don't have getpassphrase()."
+   getpassphrase_flg="#undef HAVE_GETPASSPHRASE"
+ fi
+ 
  ####### fclose
  cat > _zmachdep.c << EOF
  #include <stdio.h>
***************
*** 1840,1846 ****
  $putenv_flg
  $srand48_flg
  $srandom_flg
! $readlink_flg
  $fclose_dcl
  $pclose_dcl
  $setjmp_def
--- 1857,1863 ----
  $putenv_flg
  $srand48_flg
  $srandom_flg
! $getpassphrase_flg
  $fclose_dcl
  $pclose_dcl
  $setjmp_def
diff -cr w3m-0.2.1-inu-1.5.orig/file.c w3m-0.2.1-inu-1.5/file.c
*** w3m-0.2.1-inu-1.5.orig/file.c	Wed Oct 31 20:41:11 2001
--- w3m-0.2.1-inu-1.5/file.c	Wed Nov  7 10:50:23 2001
***************
*** 861,867 ****
  	    fflush(stdout);
  	    uname = Strfgets(stdin);
  	    Strchop(uname);
!             pwd = Strnew_charp((char *) getpass(proxy ? "Proxy Password: " : "Password: "));
  	}
  	Strcat_char(uname, ':');
  	Strcat(uname, pwd);
--- 861,879 ----
  	    fflush(stdout);
  	    uname = Strfgets(stdin);
  	    Strchop(uname);
! 	    printf(proxy ? "Proxy Password: " : "Password: ");
! 	    fflush(stdout);
! 	    if (!isatty(0)) {
! 		pwd  = Strfgets(stdin);
! 		Strchop(pwd);
! 	    }
! 	    else {
! #ifdef HAVE_GETPASSPHRASE
! 		pwd = Strnew_charp((char *) getpassphrase(""));
! #else
! 		pwd = Strnew_charp((char *) getpass(""));
! #endif
! 	    }
  	}
  	Strcat_char(uname, ':');
  	Strcat(uname, pwd);
