--- configure.orig	Fri Mar 23 11:18:40 2001
+++ configure	Mon Apr 23 15:58:47 2001
@@ -1325,6 +1325,16 @@
 	echo "Do you have directory interface? I can't find one but I hope sys/dir.h works..."
 fi
 
+####### dirent/direct
+set_binary_if='#undef USE_SET_BINARY'
+if [ $sysname = "CYGWIN" ]; then
+	echo "Your OS is CYGWIN; using setmode(, O_BINARY)"
+	set_binary_if='#define USE_SET_BINARY'
+elif [ $sysname = "OS/2" ]; then
+	echo "Your OS is OS/2; using setmode(, O_BINARY)"
+	set_binary_if='#define USE_SET_BINARY'
+fi
+
 # check signal handler
 
 do_sigtest int int
@@ -1385,6 +1395,23 @@
   no_float_h='#define NO_FLOAT_H 1'
 fi
 
+# check for srand48, lrand48
+cat > _zmachdep.c << EOF
+#include <stdlib.h>
+main()
+{
+  lrand48();
+}
+EOF
+if $cc $cflags -o _zmachdep _zmachdep.c > /dev/null 2>&1
+then
+  echo "You have srand48, lrand48."
+  no_rand48='#undef NO_RAND48'
+else
+  echo "You don't have srand48, lrand48."
+  no_rand48='#define NO_RAND48'
+fi
+
 ###### IPv6 support check
 cat > _zmachdep.c <<EOF
 #include <sys/types.h>
@@ -1648,6 +1675,7 @@
 
 $term_if
 $dir_if
+$set_binary_if
 $strcasecmp_flg
 $strchr_flg
 $strerror_flg
@@ -1684,6 +1712,7 @@
 #define NEW_FORM 1
 #define MATRIX 1
 $no_float_h
+$no_rand48
 
 #endif /* makefile_parameter */
 #endif /* _CONFIGURED_ */
--- main.c.orig	Fri Mar 23 11:14:58 2001
+++ main.c	Mon Apr 23 15:48:19 2001
@@ -59,8 +59,10 @@
 int w3m_dump = 0;
 int w3m_dump_source = 0;
 int w3m_dump_head = 0;
+int w3m_dump_extra = 0;
 static void dump_source(Buffer *);
 static void dump_head(Buffer *);
+static void dump_extra(Buffer *);
 int prec_num = 0;
 int prev_key = -1;
 int on_target = 1;
@@ -187,6 +189,29 @@
     fprintf(stderr, msg, (unsigned long)arg);
 }
 
+#ifdef NO_RAND48
+static unsigned long R1 = 0x1234abcd;
+static unsigned long R2 = 0x330e;
+#define A1 0x5deec
+#define A2 0xe66d
+#define C 0xb
+
+void
+srand48(long seed)
+{
+    R1 = (unsigned long)seed;
+    R2 = 0x330e;
+}
+
+long
+lrand48(void)
+{
+    R1 = (A1 * R1 << 16) + A1 * R2 + A2 * R1 + ((A2 * R2 + C) >> 16);
+    R2 = (A2 * R2 + C) & 0xffff;
+    return (long)(R1 >> 1);
+}
+#endif /* NO_RAND48 */
+
 int
 MAIN(int argc, char **argv, char **envp)
 {
@@ -234,16 +259,67 @@
 	    }
 	    else if (!strcmp("-h", argv[i]))
 		help();
+	    else if (!strcmp("-dump", argv[i])) {
+		argv[i] = "-dummy";
+		w3m_dump = TRUE;
+		w3m_dump_source = FALSE;
+		w3m_dump_head = FALSE;
+		w3m_halfdump = FALSE;
+		if (COLS == 0)
+		    COLS = 80;
+	    }
+	    else if (!strcmp("-dump_source", argv[i])) {
+		argv[i] = "-dummy";
+		w3m_dump = TRUE;
+		w3m_dump_source = TRUE;
+		w3m_dump_head = FALSE;
+		w3m_halfdump = FALSE;
+		if (COLS == 0)
+		    COLS = 80;
+	    }
+	    else if (!strcmp("-dump_head", argv[i])) {
+		argv[i] = "-dummy";
+		w3m_dump = TRUE;
+		w3m_dump_head = TRUE;
+		w3m_dump_source = FALSE;
+		w3m_halfdump = FALSE;
+		if (COLS == 0)
+		    COLS = 80;
+	    }
+	    else if (!strcmp("-dump_both", argv[i])) {
+		argv[i] = "-dummy";
+		w3m_dump = TRUE;
+		w3m_dump_head = TRUE;
+		w3m_dump_source = TRUE;
+		w3m_halfdump = FALSE;
+		if (COLS == 0)
+		    COLS = 80;
+	    }
+	    else if (!strcmp("-dump_extra", argv[i])) {
+#ifdef USE_SET_BINARY
+	        setmode(fileno(stdout), O_BINARY);   
+#endif
+		argv[i] = "-dummy";
+		w3m_dump = TRUE;
+		w3m_dump_head = TRUE;
+		w3m_dump_source = TRUE;
+		w3m_dump_extra = TRUE;
+		w3m_halfdump = FALSE;
+		if (COLS == 0)
+		    COLS = 80;
+	    }
+	    else if (!strcmp("-halfdump", argv[i])) {
+		argv[i] = "-dummy";
+		w3m_halfdump = TRUE;
+		w3m_dump = FALSE;
+		if (COLS == 0)
+		    COLS = 80;
+	    }
 	}
     }
 
     /* initializations */
     init_rc(config_file);
-    initKeymap();
-#ifdef MENU
-    initMenu();
-    CurrentMenuData = NULL;
-#endif				/* MENU */
 #ifdef USE_COOKIE
     initCookie();
 #endif				/* USE_COOKIE */
@@ -367,36 +443,6 @@
 		    WrapSearch = TRUE;
 		}
 	    }
-	    else if (!strcmp("-dump", argv[i])) {
-		w3m_dump = TRUE;
-		w3m_dump_source = FALSE;
-		w3m_dump_head = FALSE;
-		w3m_halfdump = FALSE;
-		if (COLS == 0)
-		    COLS = 80;
-	    }
-	    else if (!strcmp("-dump_source", argv[i])) {
-		w3m_dump = TRUE;
-		w3m_dump_source = TRUE;
-		w3m_dump_head = FALSE;
-		w3m_halfdump = FALSE;
-		if (COLS == 0)
-		    COLS = 80;
-	    }
-	    else if (!strcmp("-dump_head", argv[i])) {
-		w3m_dump = TRUE;
-		w3m_dump_head = TRUE;
-		w3m_dump_source = FALSE;
-		w3m_halfdump = FALSE;
-		if (COLS == 0)
-		    COLS = 80;
-	    }
-	    else if (!strcmp("-halfdump", argv[i])) {
-		w3m_halfdump = TRUE;
-		w3m_dump = FALSE;
-		if (COLS == 0)
-		    COLS = 80;
-	    }
 	    else if (!strcmp("-halfload", argv[i])) {
 		w3m_halfload = TRUE;
 		w3m_dump = FALSE;
@@ -526,13 +572,20 @@
     }
     if (w3m_backend)
 	backend();
-    if (!w3m_dump && !w3m_halfdump)
+    if (!w3m_dump && !w3m_halfdump) {
+	initKeymap();
+#ifdef MENU
+	initMenu();
+	CurrentMenuData = NULL;
+#endif				/* MENU */
 	fmInit();
+    }
     orig_GC_warn_proc = GC_set_warn_proc(wrap_GC_warn_proc);
     if (w3m_halfdump)
 	printf("<pre>\n");
 
     if (load_argc == 0) {
+	/* no URL specified */
 	if (!isatty(0)) {
 	    redin = newFileStream(fdopen(dup(0), "rb"), (void (*)()) pclose);
 	    newbuf = openGeneralPagerBuffer(redin);
@@ -594,7 +647,7 @@
     }
     for (; i < load_argc; i++) {
 	if (i >= 0) {
-	    if (w3m_dump && w3m_dump_head) {
+	    if (w3m_dump && w3m_dump_head && !w3m_dump_source) {
 		request = New(FormList);
 		request->method = FORM_METHOD_HEAD;
 		newbuf = loadGeneralFile(load_argv[i], NULL, NO_REFERER, 0, request);
@@ -632,23 +685,6 @@
 	    deleteFiles();
 	    exit(0);
 	}
-	if (w3m_dump) {
-	    Currentbuf = Firstbuf = newbuf;
-	    if (w3m_dump_source)
-		dump_source(Currentbuf);
-	    else if (w3m_dump_head)
-		dump_head(Currentbuf);
-	    else {
-		if (Currentbuf->frameset != NULL && RenderFrame)
-		    rFrame();
-		saveBuffer(Currentbuf, stdout);
-	    }
-	    deleteFiles();
-#ifdef USE_COOKIES
-	    save_cookies();
-#endif				/* USE_COOKIES */
-	    exit(0);
-	}
 	if (Currentbuf == NULL)
 	    Firstbuf = Currentbuf = newbuf;
 	else {
@@ -660,11 +696,33 @@
 	    rFrame();
 	    Currentbuf = newbuf;
 	}
+	if (w3m_dump) {
+	    if (w3m_dump_extra) {
+		dump_extra(Currentbuf);
+	    }
+	    if (w3m_dump_head)
+		dump_head(Currentbuf);
+	    if (w3m_dump_source)
+		dump_source(Currentbuf);
+	    if (!w3m_dump_head && !w3m_dump_source) {
+		if (Currentbuf->frameset != NULL && RenderFrame)
+		    rFrame();
+		saveBuffer(Currentbuf, stdout);
+	    }
+	}
 #ifdef BUFINFO
 	saveBufferInfo();
 #endif
 
     }
+    if (w3m_dump) {
+	deleteFiles();
+#ifdef USE_COOKIES
+	save_cookies();
+#endif				/* USE_COOKIES */
+	exit(0);
+    }
+
     if (!Firstbuf || Firstbuf == NO_BUFFER) {
 	if (newbuf == NO_BUFFER) {
 	    if (fmInitialized)
@@ -776,6 +834,17 @@
 	printf("%s", ti->ptr);
     }
     puts("");
+}
+
+static void
+dump_extra(Buffer * buf)
+{
+    printf("W3m-current-url: %s\n",parsedURL2Str(&buf->currentURL)->ptr);
+    if (buf->baseURL)
+	printf("W3m-base-url: %s\n",parsedURL2Str(buf->baseURL)->ptr);
+#ifdef JP_CHARSET
+    printf("W3m-document-charset: %s\n", code_to_str(buf->document_code));
+#endif
 }
 
 void
